### PR TITLE
Release the sampler when the active context's sampler params change (fixes #2080)

### LINKED ISSUE #2080: [Bug] LlmLiteRtCompiledModelExecutor::InitializeSampler ignores session-level sampler params (seed / temperature / topK / topP) on GPU and NPU
## Summary

`SamplerParameters` passed via `SessionConfig::sampler_params` are silently dropped on the **GPU** and **NPU** executor paths. Two same-input prompts with identical text but different `seed` produce identical output; two prompts with the same `seed` and `temperature=1.0` also produce identical output across runs. The user-facing API accepts `seed`, `temperature`, `top_k`, `top_p`, but the executor uses a hardcoded set of defaults regardless.

The CPU executor path is unaffected — `seed` / `temperature` / `top_k` / `top_p` already reach the CPU sampler correctly.

## Where the params are dropped

`runtime/executor/llm_litert_compiled_model_executor.cc:1271-1276` (commit `5e0d86b`, also in `v0.10.2`):

```cpp
proto::SamplerParameters sampler_params;
sampler_params.set_type(proto::SamplerParameters::TOP_P);
sampler_params.set_k(1);
sampler_params.set_p(0.0f);
sampler_params.set_temperature(1.0f);
sampler_params.set_seed(0);
ASSIGN_OR_RETURN(
    sampler_,
    CreateSampler(sampler_backend, output_heads, std::move(sampler_params),
                  env_.Get(), /*sequence_size=*/1, vocab_size, data_type));
```

`InitializeSampler()` constructs `SamplerParameters` from a hardcoded literal block. There is no code path that reads `session_config.sampler_params` for GPU/NPU executors. The user-supplied values reach `SessionBasic` but are forwarded only to the CPU sampler at `runtime/core/session_basic.cc:99-114`.

The closing commit on issue #1992 (`7ef9fee`) widens the Python wrapper to accept these params at the API layer but does not modify the executor, so upstream Python is also seed-deaf on GPU.

This is independent of #2073 (sampler exports gap on macOS/Windows). Even when the dynamic GPU sampler dlopen fails and `sampler_factory` falls through to the CPU sampling path *inside the executor*, the params it inherits are the same hardcoded defaults — so the gap is a symptom, not the root cause.

## Reproduction (Python, public API)

```python
import litert_lm

with litert_lm.Engine(
    "gemma-4-E2B-it.litertlm",
    backend=litert_lm.Backend.GPU,
    max_num_tokens=4096,
) as engine:

    def run_with_seed(seed: int) -> str:
        with engine.create_conversation(
            sampler_config=litert_lm.SamplerConfig(
                top_k=50,
                top_p=0.95,
                temperature=1.0,
                seed=seed,
            ),
        ) as conversation:
            message = conversation.send_message(
                "Tell a 30-word creative story about a dragon."
            )
            return message["content"][0]["text"]

    out_42a = run_with_seed(42)
    out_42b = run_with_seed(42)
    out_99  = run_with_seed(99)

    # Expected: out_42a == out_42b (same seed → identical output)
    #           out_42a != out_99   (different seed → different output)
    # Actual on GPU: all three bit-identical — every conversation uses
    #                the hardcoded defaults regardless of sampler_config.
    # On CPU (backend=Backend.CPU) the assertions hold as expected.
```

The same shape reproduces on every executor that derives from `LlmLiteRtCompiledModelExecutorBase` — Metal (macOS, iOS), WebGPU/DX12 (Windows), WebGPU/Vulkan (Linux), OpenCL (Android arm64). CPU executor on the same model honors the params correctly.

## Why a workaround inside the runtime won't help

A few options that look attractive but don't actually fix this:

1. **Plumb params through `Sampler::UpdateConfig`** — `top_p_cpu_sampler.cc:168` ignores the `seed` field; only `k`, `p`, and `temperature` are mutated. The `std::default_random_engine` is left untouched, so two consecutive `UpdateConfig` calls with different seeds still share the accumulated RNG state and produce different output for the same nominal seed. GPU C-API sampler libs (Metal, #1990) ship without `UpdateConfig` exported on some platforms.
2. **Pass via `AdvancedSettings`** — wrong scope; sampler params are session-level, not engine-level.
3. **Reuse `UpdateExecutorSettings`** — overlaps with existing semantics.
4. **Thread-local stash inside `Sampler::Create`** — decode runs on `worker_thread_pool_("engine", max=1)` (`runtime/core/engine_impl.cc:404`); TLS on the caller thread isn't visible to the worker.

The minimum-surface fix that actually works is to push session params from `SessionBasic` into the executor *before* `InitializeSampler()` runs, then have `InitializeSampler()` consume them. Happy to send a PR — I have a working patch we've been running downstream; sketch below.

## Suggested fix shape

1. Add a new virtual on `LlmExecutor`:

   ```cpp
   virtual absl::Status SetPendingSamplerParams(
       const proto::SamplerParameters& sampler_params) {
     return absl::UnimplementedError(
         "SetPendingSamplerParams not implemented for this executor.");
   }
   ```

   Default `UnimplementedError` body keeps NPU and any custom executor at current behavior.

2. Override in `LlmLiteRtCompiledModelExecutorBase`:

   ```cpp
   absl::Status SetPendingSamplerParams(
       const proto::SamplerParameters& sampler_params) override {
     pending_sampler_params_ = sampler_params;
     sampler_.reset();  // force re-create on next InitializeSampler()
     return absl::OkStatus();
   }
   ```

   Stash `std::optional<proto::SamplerParameters> pending_sampler_params_` as a member.

3. Modify `InitializeSampler()` to consume it (with backfill to today's hardcoded values for any field the caller didn't supply, so any caller that doesn't push session params is bit-identical to current behavior).

4. In `session_basic.cc` GPU/NPU branch, push `session_config.GetSamplerParams()` into the executor before `InitializeSampler()` runs. NPU returns Unimplemented and we swallow it.

This is a strict superset of current behavior — no existing caller sees a regression.

## Validation

We've been running the patch via a build-time source patcher on top of `5e0d86b` and validated end-to-end on five platforms with a regression test that runs two same-seed sessions and two different-seed sessions:

- macOS arm64 (Metal): pass (CPU + GPU)
- iOS arm64 (iPhone 16 Pro, iOS 18.7.7, Metal): pass (CPU + GPU)
- Linux x86_64 (NVIDIA L4, WebGPU/Vulkan): pass (CPU + GPU)
- Android arm64 (Pixel 8, OpenCL): pass (CPU + GPU full regression suite)
- Windows x86_64 (NVIDIA T4, WebGPU/DX12): pass (CPU + GPU)

Without the patch, GPU output is identical for `seed=42` vs `seed=99` on all five platforms. With the patch, output diverges as expected.

## Environment

- LiteRT-LM commit: `5e0d86b` (also reproduces on `v0.10.2`)
- Models tested: Gemma 4 E2B, Gemma 3 1B, Qwen3 0.6B
- Backends: WebGPU (Vulkan / DX12 / Metal), OpenCL, CPU
- Reproduction context: building `cc_binary(linkshared=True)` from `c/BUILD`, calling C API from Dart FFI in [flutter_gemma](https://github.com/DenisovAV/flutter_gemma/tree/feature/desktop-ffi)

Happy to send a PR with the patch + a reproducer test if it's useful, or to adapt the shape if a different injection point is preferred.

### PR DESCRIPTION
**Pull request by DenisovAV**
Closes #2080.

## Summary

`LlmLiteRtCompiledModelExecutorBase` builds its sampler once, from the first context's `SamplerParameters`, and `InitializeSampler()` returns early for the executor's whole lifetime. A second session on the same engine with a different `seed` / `top_k` / `top_p` / `temperature` therefore keeps sampling with the first session's settings: a default greedy first session makes every later stochastic session greedy, and (measured on v0.16.1, see #2080) the reverse as well. Still reproduces on v0.17.0.

The sampler is now released when what it was built from stops matching what `InitializeSampler()` would build — checked on a context restore (`RestoreContext()`), a runtime config update (`UpdateRuntimeConfig()`) and an executor settings update (`UpdateExecutorSettings()`) — so the next `InitializeSampler()` builds one for the context and settings actually in effect. `InitializeSampler()` itself stays build-only, and `SampleLogits()` is untouched.

## Why not at sampling time

An earlier revision of this PR detected the change inside `SampleLogits()` and rebuilt there. That is too late for a sampler that handles input: it is bound into the decode graph, and `DecodeInternal()` reads its presence as proof that this step already ran. On a restore that decodes **without** a new prefill, the first decode would hand back logits the previous context's graph produced, and they would then be sampled with the new session's configuration; the rebuilt sampler would only take effect from the following step. A restore followed by a prefill never exposed this, because `PrepareFirstPrefillAfterDecode()` disconnects a handling sampler.

For the same reason, the release disconnects the sampler from the decode graph — the same `SetSamplerInputHandling(/*reset=*/true)` used between turns — before dropping it.

## What is kept

* What an internally built sampler was constructed from is recorded in one struct: the context's sampler parameters and output-head count (its batch size, which it validates its inputs against), and the settings' resolved sampler backend and requested input-handling mode. A change that leaves all of them as they were keeps the sampler: an ordinary second turn neither rebuilds it nor restarts its random stream, and a settings update that touches none of them (the Metal residency set, for one) does not rebuild it for nothing.
* A sampler installed from outside (tests, subclasses) records no parameters and is left alone.
* The defaults each of those falls back to live in one helper apiece (`EffectiveSamplerParams`, `EffectiveOutputHeads`, `RequestedSamplerHandlesInput`), so the sampler that gets built and the comparison that decides to keep it cannot drift apart.
* The comparison includes `backend`: this executor picks the sampler backend from the engine settings, but the whole proto is forwarded to the GPU C-API samplers, so a change there is not this code's to declare irrelevant.

## Known trade-off, and a loose end

Rebuilding restarts the sampler's random stream: the seed is a construction parameter of the CPU sampler, and `UpdateConfig()` only replaces the generator when it is handed a new one. Two sessions with different parameters alternating turns therefore each restart their own stream at every switch.

`RuntimeState::rand_gen` — a per-context generator, seeded from that context's own seed in `CreateNewContext()` — looks like the intended answer, and `Sampler::UpdateConfig()` already takes one. But nothing in the runtime passes it to a sampler today: `UpdateConfig()` has no call site outside the sampler implementations themselves. Wiring it through would remove the restart and is a separate change; happy to follow up if that is the direction you want.

## Out of scope

The NPU executor samples greedily on its own path, and the MTP drafter creates its own fixed greedy samplers (k=1, p=0, temperature=1, seed=0). Neither follows per-session sampler parameters, before or after this change.

## Tests

`//runtime/executor:llm_litert_compiled_model_executor_test`, six new cases: the parameters following the active session across three restores; the release happening at restore time for a sampler bound into the decode graph, with its disconnect preceding its destruction; an output-head change in both directions, with an unset count matching an explicit one; executor settings changing the sampler backend or the input-handling mode, while an unrelated setting keeps the sampler; a restore that keeps everything keeping it, and `UpdateRuntimeConfig()` releasing it; and an externally installed sampler surviving a context switch. Dropping any of the three release calls, or any field from the comparison, fails at least one of them.

### CODE REVIEW

**Comment by DenisovAV** [comment]
I have reworked this and would appreciate a look — it has been open since April, and the bug it fixes still reproduces on v0.17.0 (`llm_litert_compiled_model_executor.cc:1339` returns early for the executor's whole lifetime, so a second session's `seed` / `top_k` / `top_p` / `temperature` never reach the sampler).

What changed since the last revision: the sampler is released where the active context changes (`RestoreContext()` / `UpdateRuntimeConfig()`) rather than when logits are sampled. The earlier placement was wrong for a sampler that handles input — it is bound into the decode graph, and `DecodeInternal()` reads its presence as proof that the step already ran, so a restore that decodes without a new prefill would have sampled the previous context's logits with the new session's configuration. `InitializeSampler()` is build-only again and `SampleLogits()` is untouched, so there is no longer any per-token work. The release also disconnects the sampler from the decode graph before dropping it.

The description has the full reasoning, including the one trade-off (a rebuilt CPU sampler restarts its random stream) and a loose end I would rather not fix silently here: `RuntimeState::rand_gen` is created per context from that context's seed, and `Sampler::UpdateConfig()` takes exactly such a generator, but nothing in the runtime ever passes it — `UpdateConfig()` has no call site outside the sampler implementations. Wiring that through would remove the restart; happy to do it in a follow-up if you want it that way.

Rebased on current `main`. Could someone take a look, or point me at the right reviewer?
